api/v1alpha3/machinehealthcheck_types.go

```diff
@@ -0,0 +1,124 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha3
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+// ANCHOR: MachineHealthCheckSpec
+
+// MachineHealthCheckSpec defines the desired state of MachineHealthCheck
+type MachineHealthCheckSpec struct {
+	// ClusterName is the name of the Cluster this object belongs to.
+	// +kubebuilder:validation:MinLength=1
+	ClusterName string `json:"clusterName"`
+
+	// Label selector to match machines whose health will be exercised
+	Selector metav1.LabelSelector `json:"selector"`
+
+	// UnhealthyConditions contains a list of the conditions that determine
+	// whether a node is considered unhealthy.  The conditions are combined in a
+	// logical OR, i.e. if any of the conditions is met, the node is unhealthy.
+	//
+	// +kubebuilder:validation:MinItems=1
+	UnhealthyConditions []UnhealthyCondition `json:"unhealthyConditions"`
+
+	// Any farther remediation is only allowed if at most "MaxUnhealthy" machines selected by
+	// "selector" are not healthy.
+	// +optional
+	MaxUnhealthy *intstr.IntOrString `json:"maxUnhealthy,omitempty"`
+}
+
+// ANCHOR_END: MachineHealthCHeckSpec
+
+// ANCHOR: UnhealthyCondition
+
+// UnhealthyCondition represents a Node condition type and value with a timeout
+// specified as a duration.  When the named condition has been in the given
+// status for at least the timeout value, a node is considered unhealthy.
+type UnhealthyCondition struct {
+	// +kubebuilder:validation:Type=string
+	// +kubebuilder:validation:MinLength=1
+	Type corev1.NodeConditionType `json:"type"`
+
+	// +kubebuilder:validation:Type=string
+	// +kubebuilder:validation:MinLength=1
+	Status corev1.ConditionStatus `json:"status"`
+
+	// It would be preferable for timeout to be a metav1.Duration, but there's
+	// no good way to validate the format here.  Invalid input would cause
+	// problems with marshaling, so it's better to just make it a string and
+	// handle the conversion in the controller.
```

Comment:
Why couldn't it be validated through a validating webhook?

Reply:
I'm porting this project from our openshift implementation, at the time the comment was written we didn't have support for webhooks because of how early we forked cluster-api to make our implementation.

I believe you're right, this could totally be validated by a webhook! Does cluster-api already have some validating webhooks? Is this something I should be adding to the PR do you think?

Reply:
@JoelSpeed yes, we are currently using webhooks for defaulting, validation, and conversion in other places. The defaulting/validation for Cluster resources can be found here: https://github.com/kubernetes-sigs/cluster-api/blob/ee0b0eb29a571aa72894abb79729dccc1f75caca/api/v1alpha3/cluster_webhook.go

Reply:
@detiber I've added a webhook to validate the durations and the selectors that we have on the spec.

I've left the type itself as a string so that we don't get marshalling errors, which I believe would give less helpful descriptions of the problem, and updated the comment to reflect this

Reply:
lol, it looks like it's a TODO in controller-tools to better handle metav1.Duration through openapi validation :Joy: https://github.com/kubernetes-sigs/controller-tools/blob/f5dffbacb9403d1e6d3ba68634d89bb8724d1c13/pkg/crd/known_types.go#L43

Reply:
Haha, maybe they will fix that by v1alpha4 🤷‍♂ Are we agreed it's best to leave it as a string for now?

Reply:
I went ahead and started an attempt to add support for it in controller-tools (https://github.com/kubernetes-sigs/controller-tools/pull/394), I'll do some additional testing on it, but we could apply a similar validation by overriding with kubebuilder validation annotations as an interim step.

Reply:
For kubebuilder validation in the interim, that would be applying the same regex (from your PR) as a `// +kubebuilder:validation:Pattern` right? This can only be used on a string otherwise the controller-gen throws an error back at you, so even with that, we would have to keep the type as a `string` until your controller-tools PR is merged as far as I understand

Reply:
@JoelSpeed I ended up closing the controller-tools PR, since pattern matching is happening after the resource is unmarshalled.

This brings up a bigger point wrt how we are handling other resources (such as anywhere we are exposing metav1.Time, or the embedded Kubeadm types which contain metav1.Duration fields, which also present the same challenge).

Personally, I don't find the unmarshalling error that bad, and it appears to be coming from the defaulting webhook, so I'm wondering if that might be the proper place to clean up these types of errors?

Reply:
@ncdc @vincepri thoughts?

Reply:
An example of the unmarshalling error here:
```
Error from server: error when creating "/home/detiber/tmp.txt": admission webhook "default.kubeadmcontrolplane.controlplane.cluster.x-k8s.io" denied the request: v1alpha3.KubeadmControlPlane.Spec: v1alpha3.KubeadmControlPlaneSpec.Version: UpgradeAfter: unmarshalerDecoder: parsing time "blueberry" as "2006-01-02T15:04:05Z07:00": cannot parse "blueberry" as "2006", error found in #10 byte of ...|blueberry","version"|..., bigger context ...|ation":null},"test":"0","upgradeAfter":"blueberry","version":"v1.16.2"}}|...
```

Reply:
I think I'm ok with the unmarshalling error until we can get better error handling in controller-runtime. I'd rather have the data type and OpenAPI schema match what we expect (Duration, not string).

Reply:
I've changed it to a `metav1.Duration`